File: src/db-gwent-server-scala/src/main/scala/entities/Game.scala
package entities

case class GameId(value: String)
case class Game(id: GameId, players: Vector[Player], turn: Turn):
  def this(id: GameId, players: Vector[Player]) =
    this(id, players, Turn(players.map(p => p.id).toList, PlayerId("")))

  def playCard(playerId: PlayerId, cardId: CardId, row: TargetRowType = null): Boolean =
    turn.run(playerId, () => players.find(p => p.id == playerId).get.playCard(cardId, row))

File: src/db-gwent-server-kotlin/src/main/kotlin/com/entities/Game.kt
package com.entities

class Game(
    val id: GameId,
    val players: List<Player>,
    val turn: Turn = Turn(players.map{ it.id }.toList())) {

    fun playCard(playerId: PlayerId, cardId: CardId, row: RowType? = null): Boolean =
        turn.run(playerId) {
            val thisPlayer = players.find { it.id == playerId } ?: return@run false
            val nextPlayer = players.find { it.id != playerId } ?: return@run false

            thisPlayer.playCard(cardId, row, nextPlayer)

            val card = thisPlayer.getCard(cardId) ?: return@run false
            card.abilities.all { it.apply(card, thisPlayer, nextPlayer) }
        }

}

data class GameId(val value: String)
